File: image_switch/js/image_switch.js
import { app } from "../../scripts/app.js";

const MAX_INPUTS = 32;

app.registerExtension({
    name: "ImageSwitch",

    async beforeRegisterNodeDef(nodeType, nodeData) {
        if (nodeData.name !== "ImageSwitch") return;

        // --- Node creation ---
        const onNodeCreated = nodeType.prototype.onNodeCreated;
        nodeType.prototype.onNodeCreated = function () {
            onNodeCreated?.apply(this, arguments);

            // Hide the upstream_titles widget — internal plumbing only
            const titlesWidget = this.widgets.find(w => w.name === "upstream_titles");
            if (titlesWidget) {
                titlesWidget.type = "hidden";
                titlesWidget.computeSize = () => [0, -4];
            }

            this.addWidget("button", "Update Inputs", null, () => {
                const numWidget  = this.widgets.find(w => w.name === "num_inputs");
                const maskWidget = this.widgets.find(w => w.name === "use_masks");
                if (!numWidget) return;
                updateInputSockets(this, numWidget.value, maskWidget?.value ?? false);
            }, { serialize: false });

            const numWidget  = this.widgets.find(w => w.name === "num_inputs");
            const maskWidget = this.widgets.find(w => w.name === "use_masks");
            updateInputSockets(this, numWidget?.value ?? 2, maskWidget?.value ?? false);
        };

        // --- Serialization: always snapshot all upstream titles before prompt is sent ---
        const onSerialize = nodeType.prototype.onSerialize;
        nodeType.prototype.onSerialize = function (data) {
            onSerialize?.apply(this, arguments);
            writeAllUpstreamTitles(this);
            data.num_inputs = this.widgets.find(w => w.name === "num_inputs")?.value ?? 2;
            data.use_masks  = this.widgets.find(w => w.name === "use_masks")?.value  ?? false;
        };

        // --- Deserialization ---
        const onConfigure = nodeType.prototype.onConfigure;
        nodeType.prototype.onConfigure = function (data) {
            onConfigure?.apply(this, arguments);
            const count    = data.num_inputs ?? 2;
            const useMasks = data.use_masks  ?? false;

            const numWidget  = this.widgets.find(w => w.name === "num_inputs");
            const maskWidget = this.widgets.find(w => w.name === "use_masks");
            if (numWidget)  numWidget.value  = count;
            if (maskWidget) maskWidget.value = useMasks;

            updateInputSockets(this, count, useMasks);
        };
    },

    // Last chance before prompt is queued — covers the socket-driven 'which' case
    // by snapshotting ALL input titles; Python selects the right one at runtime
    beforeQueuePrompt() {
        for (const node of app.graph._nodes) {
            if (node.type === "ImageSwitch") {
                writeAllUpstreamTitles(node);
            }
        }
    },
});

/**
 * Walk every image_N input, resolve its upstream node title, and write the
 * full array as JSON into the hidden upstream_titles widget.
 * Python picks the correct entry using the runtime value of 'which'.
 */
function writeAllUpstreamTitles(node) {
    const titlesWidget = node.widgets?.find(w => w.name === "upstream_titles");
    const numWidget    = node.widgets?.find(w => w.name === "num_inputs");
    if (!titlesWidget) return;

    const count = numWidget?.value ?? MAX_INPUTS;
    const titles = [];

    for (let i = 0; i < count; i++) {
        const inputSlot = node.inputs?.findIndex(inp => inp.name === `image_${i}`);
        if (inputSlot == null || inputSlot === -1) {
            titles.push("");
            continue;
        }

        const linkId = node.inputs[inputSlot].link;
        if (linkId == null) {
            titles.push("");
            continue;
        }

        const link = node.graph.links[linkId];
        if (!link) {
            titles.push("");
            continue;
        }

        const upstreamNode = node.graph.getNodeById(link.origin_id);
        titles.push(upstreamNode?.title ?? upstreamNode?.type ?? "");
    }

    titlesWidget.value = JSON.stringify(titles);
}

/**
 * Add or remove image_N (and optionally mask_N) input sockets.
 */
function updateInputSockets(node, count, useMasks) {
    count = Math.max(1, Math.min(MAX_INPUTS, count));

    const existingInputs = node.inputs ? [...node.inputs] : [];
    const imageSlots = existingInputs.filter(i => i.name.startsWith("image_")).length;
    const maskSlots  = existingInputs.filter(i => i.name.startsWith("mask_")).length;

    // Reconcile image sockets
    if (count > imageSlots) {
        for (let i = imageSlots; i < count; i++) node.addInput(`image_${i}`, "IMAGE");
    } else if (count < imageSlots) {
        for (let i = imageSlots - 1; i >= count; i--) removeNamedInput(node, `image_${i}`);
    }

    // Reconcile mask sockets
    if (useMasks) {
        if (count > maskSlots) {
            for (let i = maskSlots; i < count; i++) node.addInput(`mask_${i}`, "MASK");
        } else if (count < maskSlots) {
            for (let i = maskSlots - 1; i >= count; i--) removeNamedInput(node, `mask_${i}`);
        }
    } else {
        for (let i = maskSlots - 1; i >= 0; i--) removeNamedInput(node, `mask_${i}`);
    }

    // Clamp 'which' max to count - 1
    const whichWidget = node.widgets?.find(w => w.name === "which");
    if (whichWidget) {
        whichWidget.options.max = count - 1;
        if (whichWidget.value >= count) whichWidget.value = count - 1;
    }

    node.setDirtyCanvas(true, true);
}

function removeNamedInput(node, name) {
    const idx = node.inputs?.findIndex(i => i.name === name);
    if (idx == null || idx === -1) return;
    if (node.inputs[idx].link != null) node.graph.removeLink(node.inputs[idx].link);
    node.removeInput(idx);
}

File: image_switch/image_switch.py
"""
ImageSwitch - A Nuke-style Switch node for ComfyUI
Connects multiple image inputs and passes through the one selected by 'which'.
- 'which' is zero-indexed (0 = first input)
- 'which' widget can be converted to an INT input socket
- 'num_inputs' controls how many image sockets are shown (updated via JS button)
- 'use_masks' enables a corresponding mask_N socket for each image_N
- Outputs IMAGE, MASK, and STRING (title of the upstream node on the selected input)
- JS serializes all upstream titles as JSON; Python picks the active one by 'which'
- mask output is None if not connected; string output is "" if not connected
"""

import json

MAX_INPUTS = 32


class ImageSwitch:
    @classmethod
    def INPUT_TYPES(cls):
        optional = {}
        for i in range(MAX_INPUTS):
            optional[f"image_{i}"] = ("IMAGE",)
            optional[f"mask_{i}"] = ("MASK",)
        return {
            "required": {
                "which": ("INT", {
                    "default": 0,
                    "min": 0,
                    "max": MAX_INPUTS - 1,
                    "step": 1,
                    "display": "number"
                }),
                "num_inputs": ("INT", {
                    "default": 2,
                    "min": 1,
                    "max": MAX_INPUTS,
                    "step": 1,
                    "display": "number"
                }),
                "use_masks": ("BOOLEAN", {"default": False}),
                # Written by JS before execution — JSON array of all upstream titles
                "upstream_titles": ("STRING", {"default": "[]"}),
            },
            "optional": optional,
        }

    RETURN_TYPES = ("IMAGE", "MASK", "STRING")
    RETURN_NAMES = ("image", "mask", "node_name")
    FUNCTION = "switch"
    CATEGORY = "image/utils"

    def switch(self, which, num_inputs, use_masks, upstream_titles, **kwargs):
        # Resolve image, with fallback to nearest connected input
        image = kwargs.get(f"image_{which}")
        resolved_index = which

        if image is None:
            for i in range(which - 1, -1, -1):
                fallback = kwargs.get(f"image_{i}")
                if fallback is not None:
                    print(f"[ImageSwitch] Input {which} not connected, falling back to input {i}")
                    image = fallback
                    resolved_index = i
                    break
            if image is None:
                raise ValueError(
                    f"[ImageSwitch] 'which' is set to {which} but no inputs are connected."
                )

        # Resolve mask
        mask = kwargs.get(f"mask_{resolved_index}") if use_masks else None

        # Pick the upstream title for the resolved index
        try:
            titles = json.loads(upstream_titles)
        except Exception:
            titles = []

        node_name = titles[resolved_index] if resolved_index < len(titles) else ""

        return (image, mask, node_name)


NODE_CLASS_MAPPINGS = {
    "ImageSwitch": ImageSwitch,
}

NODE_DISPLAY_NAME_MAPPINGS = {
    "ImageSwitch": "Switch/Which",
}
